src/System.Buffers/tests/ArrayPool/UnitTests.cs

```diff
@@ -371,5 +371,13 @@ public static void ExhaustingBufferBucketFiresDiagnosticEvent()
                 Assert.True(are.WaitOne(MaxEventWaitTimeoutInMs));
             }, 4, are);
         }
+
+        [Fact]
+        public static void ReturningANonPooledBufferOfDifferentSizeToThePoolThrows()
+        {
+            ArrayPool<byte> pool = ArrayPool<byte>.Create(maxArrayLength: 16, maxArraysPerBucket: 1);
+            byte[] buffer = pool.Rent(15);
+            Assert.Throws<ArgumentException>("buffer", () => pool.Return(new byte[1]));
+        }
```

Comment:
Nit: You might still add a subsequent ```pool.Rent(15)``` call and validate its length to make sure that the previous buffer wasn't returned before the exception was thrown.